### DISCUSSION TITLE
Positive and Negative Sampling using NeighborLoader

### DISCUSSION THREAD

**Discussion by wilcoln** [pyg-team/pytorch_geometric#4330] [Q&A] [answered]
Hello,
I am trying to train a GraphSAGE model in an unsupervised manner on the Reddit dataset. For this, I am taking inspiration from the [unsupervised graphsage example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup.py) to adapt the [supervised Reddit example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/reddit.py). In the former example, positive and negative sampling is achieved by appropriately extending the `NeighborSampler` class. However, `NeighborSampler` has now been deprecated, and the latter example instead makes use of `NeighborLoader`. Unfortunately, I couldn't find any example showing how to use/extend this latter class to perform positive and negative sampling. I've been struggling on this for a couple of days now. Any help will be warmly welcomed!

**Comment by wilcoln** [comment, marked as answer]
Update: I managed to train the model using the deprecated `NeighborSampler`.

**Reply by rusty1s** [reply]
That sounds good. Glad that you got it working.

I will also try to convert this example to make use of the `NeighborLoader` class. We are working on better support for this in https://github.com/pyg-team/pytorch_geometric/issues/4026.

**Comment by daviddavo** [comment]
How did you do it? I want to do BPR and I also need positive/negative sampling with NeighborLoader, but I don't know how.

**Reply by rusty1s** [reply]
Take a look at `LinkNeighborLoader` :)

**Reply by daviddavo** [reply]
I want something between NeighborLoader and LinkNeighborLoader: Something that can return a triplet (like `LinkNeighborLoader`, just what I want, yay!), but iterates through every (source) node just once, instead of returning a sample for every link.

**Reply by rusty1s** [reply]
What about initializing two `NeighborLoader` instances (the second with random seed nodes), and iterating through them via
```python
for batch1, batch2 in zip(loader1, loader2):
     pass

**Reply by daviddavo** [reply]
The seed nodes need to be the same. I need negative samples for each source node in a bipartite graph.

**Reply by rusty1s** [reply]
In this case, you can specify a `sampler` argument to both loaders to guarantee that they share the same seed nodes.
